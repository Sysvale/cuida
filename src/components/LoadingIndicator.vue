<template>
	<div :class="`loading-indicator--${variant}`" />
</template>

<script setup>
import { computed, ref, watch } from 'vue';
import isDeviceType from '../utils/methods/isDeviceType.js';

// Props
const props = defineProps({
	/**
	 * Controla a exibição do LoadingIndicator.
	 */
	modelValue: {
		type: Boolean,
		default: false,
		required: true,
	},
	/**
	 * A variante de cor. São 10 variantes implementadas: 'green', 'teal',
	 * 'blue', 'indigo', 'violet', 'pink', 'red', 'orange', 'amber' e 'turquoise'.
	 */
	variant: {
		type: String,
		default: 'green',
	},
	/**
	 * Define o tempo, em ms, do intervalo de mudança da barra de progresso.
	 */
	intervalTime: {
		type: Number,
		default: 500,
	},
});

// Variáveis reativas
const currentPercentage = ref(0);

// Variáveis computadas
const barPercentage = computed(() => {
	return `${currentPercentage.value}%`;
});

const disabledTransition = computed(() => {
	return currentPercentage.value === 0 ? 'none' : 'width 0.5s';
});

const barHeight = computed(() => {
	if (isDeviceType('smartphone') || isDeviceType('tablet')) {
		return '4px';
	}

	return '3px';
})

// Watch
watch(() => props.modelValue, (newValue) => {
	if (newValue) {
		start();
		increment();
		return;
	}

	finish();
});

// Métodos
function start() {
	currentPercentage.value = 1;
}

function increment() {
	if (currentPercentage.value > 0 && currentPercentage.value < 90) {
		currentPercentage.value += Math.floor(Math.random() * (25 - 10 + 1) + 10);

		if (currentPercentage.value >= 90) {
			currentPercentage.value = 90;
			return;
		}

		let randomInterval = Math.floor(Math.random() * ((props.intervalTime * 1.4) - (props.intervalTime * 0.5)) + (props.intervalTime * 0.5));
		setTimeout(increment, randomInterval);
	}
}

function finish() {
	currentPercentage.value = 100;
	setTimeout(() => {
		currentPercentage.value = 0;
	}, 600);
}

</script>

<style lang="scss" scoped>
@import '../assets/sass/tokens.scss';

.loading-indicator {
	position: fixed;
	top: 0;
	left: 0;
	width: v-bind(barPercentage);
	height: v-bind(barHeight);
	display: flex;
	justify-content: center;
	align-items: center;
	background-color: $n-10;
	background-size: 200% 100%;
	animation: loading-gradient 2s ease infinite;

	@keyframes loading-gradient {
		0% {
			background-position: 0% 0%;
		}
		100% {
			background-position: -100% 0%;
		}
	}
	z-index: $z-index-toast;
	transition: v-bind(disabledTransition);

	@include variantResolver using ($color-name, $shade-50, $shade-100, $shade-200, $shade-300, $base-color, $shade-500, $shade-600) {
		@extend .loading-indicator;
		background-image: linear-gradient(to right, rgba($base-color, 0.6), rgba($base-color, 1));
	}
}
</style>
