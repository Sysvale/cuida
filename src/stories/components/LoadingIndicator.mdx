import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import LoadingIndicator from '../../components/LoadingIndicator.vue';
import Button from '../../components/Button.vue';
import * as LoadingIndicatorStories from './LoadingIndicator.stories';

<Meta of={LoadingIndicatorStories} />

export const Template = (args) => ({
	setup() {
		return { args };
	},
	data() {
		return {
			showLoadingIndicator: false,
		};
	},
	components: { CdsLoadingIndicator: LoadingIndicator, CdsButton: Button },
	template: /*html*/ `
		<div class="d-flex justify-content-center">
			<cds-button
				id="trigger-popover"
				@click="showLoadingIndicator = !showLoadingIndicator"
			>
				Toggle LoadingIndicator
			</cds-button>
			<cds-loading-indicator
				v-bind="args"
				style="margin-top: 71px;"
				v-model="showLoadingIndicator"
			/>
		</div>
	`,
	updated() {
		console.info('%cv-model: ', 'color: #2C70CD;', this.showLoadingIndicator);
	},
});

# LoadingIndicator

### O LoadingIndicator é um componente utilizado para indicar o progresso do carregamento da página, proporcionando uma experiência de usuário mais fluida e informativa.

***

<br />

## Quando usar:

* Em navegações entre páginas ou abas dentro de uma aplicação, onde o usuário precisa visualizar o progresso de carregamento de forma clara para saber que a transição está em andamento;
* Em aplicações onde o carregamento de conteúdo pode demorar, como em operações de rede ou carregamento de dados pesados.

<br />

## Quando não usar:

* Para o carregamento de dados secundários ou ações que não afetam diretamente a navegação principal do usuário;
* Quando o usuário realiza interações que podem desencadear novos carregamentos, como o acionamento de um toggle switch. Nesses casos, considere o uso de alternativas como o **Skeleton** para mostrar o conteúdo sendo carregado ou o **Spinner** para indicar que uma ação está em andamento.

<br />

## Preview

<Canvas>
  <Story of={LoadingIndicatorStories.LoadingIndicator} />
</Canvas>

<Controls />
