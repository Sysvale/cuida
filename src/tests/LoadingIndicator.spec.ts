import { describe, test, expect } from 'vitest';
import LoadingIndicator from '../components/LoadingIndicator.vue';
import { mount } from '@vue/test-utils';

describe('LoadingIndicator', () => {
	test('renders correctly', async () => {
		const wrapper = mount(LoadingIndicator, {
			props: {
				modelValue: true,
				variant: 'green',
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
