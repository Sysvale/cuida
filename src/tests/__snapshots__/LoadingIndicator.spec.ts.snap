// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoadingIndicator > renders correctly 1`] = `"<div data-v-2f1c657e="" class="loading-indicator--green"></div>"`;
